JENKINS-71052: CSP compatibility: SetupWizard/wizard-ui.jelly

niralmaruda opened the pull request:
<!-- Comment:
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue ID you created in Jira.
Note that if you want your changes backported into LTS, you need to create a Jira issue. See https://www.jenkins.io/download/lts/#backporting-process for more information.
-->
## SetupWizard/wizard-ui.jelly has inline JavaScript.
See [JENKINS-71052](https://issues.jenkins.io/browse/JENKINS-71052).

<!-- Comment:
If the issue is not fully described in Jira, add more information here (justification, pull request links, etc.).

 * We do not require Jira issues for minor improvements.
 * Bug fixes should have a Jira issue to facilitate the backporting process.
 * Major new features should have a Jira issue.
-->

### Testing done
Ran at my local pc and it worked fine. Not sure about additional tests.

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

### Proposed changelog entries

- SetupWizard/wizard-ui.jelly has inline JavaScript. So I have followed new inline JavaScript guideline and updated it. 

<!-- Comment:
The changelog entry should be in the imperative mood; e.g., write "do this"/"return that" rather than "does this"/"returns that".
For examples, see: https://www.jenkins.io/changelog/
-->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] The Jira issue, if it exists, is well-described.
- [X] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [X] There is automated testing or an explanation as to why this change has no tests.
- [X] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [X] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [X] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [X] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [X] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@NotMyFault @MarkEWaite @janfaracik 
Kindly please fell free to guide me if there are any mistakes in my pull request. 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7914"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

niralmaruda committed:
JENKINS-71052: CSP compatibility: SetupWizard/wizard-ui.jelly

niralmaruda committed:
Merge branch 'jenkinsci:master' into JENKINS-71052

niralmaruda committed:
Merge branch 'jenkinsci:master' into JENKINS-71052

niralmaruda committed:
Merge branch 'jenkinsci:master' into JENKINS-71052

NotMyFault commented:
> Testing done
> Ran at my local pc and it worked fine

How did you compile these changes? `mvn clean verify -DskipTests` will show the same failures you can observe on the PR build on ci.jenkins.io

NotMyFault commented:
> I followed the same steps mentioned in docs for this bug. And yet I didn't know how can I resolve this. I understand that the variables which I have declared is not being used but I can't figure out where exactly I have to use it and how? If you could guide me where to look at. It will be really helpful.
> Thank you.

ESLint [documents](https://eslint.org/docs/latest/rules/no-unused-vars) the meaning of this error and how to suppress it in case of a false-positive.

yaroslavafenkin commented on core/src/main/resources/jenkins/install/SetupWizard/wizard-ui.jelly:
```diff
@@ -1,13 +1,7 @@
-      setupWizardExtensions.push(extension);
-    };
-  </script>
+  <div id="target-div" data-default-update-site-id="${hudson.model.UpdateCenter.ID_DEFAULT}"></div>
```
`target-div` is a bit too generic as for me, I'd suggest something more specific to avoid potential conflicts

niralmaruda replied:
Is this id name is okay? Because this div will return default site id.
`<div id="default-site-id" data-default-update-site-id="${hudson.model.UpdateCenter.ID_DEFAULT}"></div>`

yaroslavafenkin commented on core/src/main/resources/jenkins/install/SetupWizard/wizardUi.js:
```diff
@@ -0,0 +1,8 @@
+var targetDiv = document.querySelector('#target-div');
```
Not a huge deal, but variables declared here are in global scope and will be available everywhere. So maybe try avoid declaring unnecessary ones.

niralmaruda replied:
I have changed variable name to `defaultSiteId`. and later we'll update the default site id so that'll be `defaultUpdateSiteId`

`var defaultSiteId = document.querySelector('#default-site-id');`

yaroslavafenkin replied:
What I was saying is that you could inline some calls to avoid declaring extra variables.

yaroslavafenkin reviewed (commented):
I've tested it manually and it seems fine to me functionally.

As for eslint errors, `defaultUpdateSiteId` is used in https://github.com/jenkinsci/jenkins/blob/e3fdfa527e4fefb4b37f04c92a2dd87b8a374b75/war/src/main/js/pluginSetupWizardGui.js#L1323

And `onSetupWizardInitialized` is referenced from https://github.com/jenkinsci/jenkins/blob/e27b310065b3c036b5fc9d123f1d1d99d3058c00/core/src/main/java/jenkins/install/InstallState.java#L47

So both seem to be false positives.

niralmaruda commented:
@NotMyFault @yaroslavafenkin for false positive. I tried following things as mentioned in [document](https://eslint.org/docs/latest/rules/no-unused-vars) but I'm still getting same error.
```
/* exported defaultUpdateSiteId*/
```
```
/*eslint no-unused-vars: ["error", { "vars": "local" }]*/
/*global defaultUpdateSiteId*/
```
```
/*eslint no-unused-vars: ["error", { "vars": "all" }]*/
/*global defaultUpdateSiteId*/
```


And when I use
```
/*eslint no-unused-vars: ["error", { "vars": "local" }]*/
/*global setupWizardExtensions */
```
I get this error
```  
  2:10  error  'setupWizardExtensions' is already defined as a built-in global variable  no-redeclare
  6:5   error  'defaultUpdateSiteId' is assigned a value but never used                  no-unused-vars
  9:5   error  'onSetupWizardInitialized' is assigned a value but never used             no-unused-vars
```

niralmaruda committed:
Fixed wizardUi.js

`// eslint-disable-next-line no-unused-vars` to fix no-unused-vars error
`cd .\war\; yarn lint:fix` to fix prettier error.

niralmaruda committed:
Merge branch 'jenkinsci:master' into JENKINS-71052

mawinter69 commented:
> I've tested it manually and it seems fine to me functionally.
> 
> As for eslint errors, `defaultUpdateSiteId` is used in
> 
> https://github.com/jenkinsci/jenkins/blob/e3fdfa527e4fefb4b37f04c92a2dd87b8a374b75/war/src/main/js/pluginSetupWizardGui.js#L1323
> 
> And `onSetupWizardInitialized` is referenced from
> 
> https://github.com/jenkinsci/jenkins/blob/e27b310065b3c036b5fc9d123f1d1d99d3058c00/core/src/main/java/jenkins/install/InstallState.java#L47
> 
> So both seem to be false positives.

Wouldn't it make sense to move the javascript code to https://github.com/jenkinsci/jenkins/blob/e3fdfa527e4fefb4b37f04c92a2dd87b8a374b75/war/src/main/js/pluginSetupWizardGui.js directly
Then we have the setting of the variables and the consumption in one place.

yaroslavafenkin commented:
> Wouldn't it make sense to move the javascript code to https://github.com/jenkinsci/jenkins/blob/e3fdfa527e4fefb4b37f04c92a2dd87b8a374b75/war/src/main/js/pluginSetupWizardGui.js directly Then we have the setting of the variables and the consumption in one place.

Makes sense to me. Needs to be done with care though, as far as I understand `onSetupWizardInitialized` has to be available in global scope.

niralmaruda committed:
change implementation for wizardUI

Deleted Wizardui.js and added logic to pluginSetupWizardGui.js

niralmaruda committed:
Merge branch 'jenkinsci:master' into JENKINS-71052

yaroslavafenkin commented on core/src/main/resources/jenkins/install/SetupWizard/wizard-ui.jelly:
```diff
@@ -1,13 +1,7 @@
-    };
-  </script>
+  <div id="default-site-id" data-default-update-site-id="${hudson.model.UpdateCenter.ID_DEFAULT}"></div>
+  <st:adjunct includes="jenkins.install.SetupWizard.wizardUi"/>
```
This does not exist anymore

yaroslavafenkin commented on war/src/main/js/pluginSetupWizardGui.js:
```diff
@@ -1282,6 +1282,20 @@ var createPluginSetupWizard = function (appendTarget) {
   // do this so the page isn't blank while doing connectivity checks and other downloads
   setPanel(loadingPanel);
 
+  /* exported defaultUpdateSiteId */
```
There's an error now: 
![image](https://github.com/jenkinsci/jenkins/assets/91559310/533bcfb3-9154-4947-a4fb-9d468e200774)

niralmaruda replied:
@yaroslavafenkin Can you let me know how did you reproduced this error? So I can also test it

yaroslavafenkin replied:
Start a fresh instance of Jenkins, it would be the first thing you'd see after entering initial admin password.

yaroslavafenkin commented on war/src/main/js/pluginSetupWizardGui.js:
```diff
@@ -1282,6 +1282,20 @@ var createPluginSetupWizard = function (appendTarget) {
   setPanel(loadingPanel);
 
+  /* exported defaultUpdateSiteId */
+  var defaultUpdateSiteId = function () {
+    var defaultSiteId = document
+      .querySelector("#default-site-id")
+      .getAttribute("data-default-update-site-id");
+    return defaultSiteId ? defaultSiteId.replace("'", "") : "default";
+  };
```
```suggestion
  var defaultUpdateSiteId = (function () {
    var defaultSiteId = document
      .querySelector("#default-site-id")
      .getAttribute("data-default-update-site-id");
    return defaultSiteId ? defaultSiteId.replace("'", "") : "default";
  })();
```
Perhaps IIFE syntax would work here? Haven't tested though.

niralmaruda replied:
I did the changes you mentioned above and the ran `mvn clean verify -DskipTests` got successfully build message.
```
[INFO] Jenkins main module ................................ SUCCESS [  4.399 s]
[INFO] Jenkins BOM ........................................ SUCCESS [  0.291 s]
[INFO] Internal SPI for WebSocket ......................... SUCCESS [  3.260 s]
[INFO] Jetty 10 implementation for WebSocket .............. SUCCESS [  2.961 s]
[INFO] Jenkins cli ........................................ SUCCESS [  7.056 s]
[INFO] Jenkins core ....................................... SUCCESS [01:10 min]
[INFO] Jenkins war ........................................ SUCCESS [ 46.671 s]
[INFO] Tests for Jenkins core ............................. SUCCESS [ 18.275 s]
[INFO] Jenkins coverage ................................... SUCCESS [  0.465 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```
If anything else needs to be done in order to test this change please do let me know. If not, then give a thumbs up to this message so I can commit the changes.

yaroslavafenkin replied:
To test it check whether it actually works in runtime. I'd go through setup wizard steps a few times with console open and observe whether something unusual pops up.

niralmaruda replied:
@yaroslavafenkin should I install all suggested plugins or is it not needed? Or should I test with both suggested select and not selecting anything?

yaroslavafenkin replied:
Installed plugins don't matter here.

niralmaruda committed:
remove adjunct for wizardui

niralmaruda committed:
Merge branch 'jenkinsci:master' into JENKINS-71052

niralmaruda committed:
change for global scope

Changed onSetupWizardInitialized to global scope.

niralmaruda committed:
Merge branch 'jenkinsci:master' into JENKINS-71052

niralmaruda committed:
update global scope

exported onSetupWizardInitialized to global scope.

niralmaruda committed:
Merge branch 'JENKINS-71052' of https://github.com/niralmaruda/jenkins into JENKINS-71052

niralmaruda committed:
Revert "Merge branch 'JENKINS-71052' of https://github.com/niralmaruda/jenkins into JENKINS-71052"

This reverts commit b82a857e1ff3744add95b479c5a3fc7bed537748, reversing
changes made to d546ee151a67a8eccb8b553c1331b63991d829a6.

niralmaruda committed:
Revert "Revert "Merge branch 'JENKINS-71052' of https://github.com/niralmaruda/jenkins into JENKINS-71052""

This reverts commit bf8e92700c4c923755b15e76fd9ac64c045e6634.

niralmaruda committed:
Update pluginSetupWizardGui.js

niralmaruda commented on war/src/main/js/pluginSetupWizardGui.js:
```diff
@@ -1282,6 +1282,20 @@ var createPluginSetupWizard = function (appendTarget) {
+
+  var setupWizardExtensions = [];
+  // eslint-disable-next-line no-unused-vars
+  var onSetupWizardInitialized = function (extension) {
```
@basil @yaroslavafenkin I've added `/* exported onSetupWizardInitialized */`, that will make `onSetupWizardInitialized` global if I'm correct.

yaroslavafenkin replied:
I don't understand how that's supposed to make it global. I don't see any other examples of this in the repo, so this seems to be nothing more than a regular comment.

I think this should do it, but I'm not entirely sure:
```suggestion
  window.onSetupWizardInitialized = function (extension) {
```

niralmaruda replied:
Hello @yaroslavafenkin,
I can done the changes that you have suggested. Can please review it?

Thank you 😄

yaroslavafenkin replied:
Sure, the code looks fine, I'll test whether this actually puts the function into the global scope, because JS is weird sometimes.

yaroslavafenkin replied:
Hey @niralmaruda,

Sorry, it took me a while to test it.
It won't work in it's current form. Declaring `onSetupWizardInitialized` inside `createPluginSetupWizard` is a bit too late. Function is expected to exist before `createPluginSetupWizard` runs.

I suggest that you switch back to the approach you have before, i.e. having the JS code inside e.g. `_wizard-ui.js`:
```js
// all variables declared here have to be in global scope
window.defaultUpdateSiteId = (function () {
  var defaultSiteId = document
    .querySelector("#default-site-id")
    .getAttribute("data-default-update-site-id");
  return defaultSiteId ? defaultSiteId.replace("'", "") : "default";
})();

window.setupWizardExtensions = [];
window.onSetupWizardInitialized = function (extension) {
  setupWizardExtensions.push(extension);
};
```
and have it `st:adjunct` ed in https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/jenkins/install/SetupWizard/wizard-ui.jelly

niralmaruda committed:
add global windows object.

Expose the functions to the window objects so it can be used globally.

niralmaruda committed:
Merge branch 'master' into JENKINS-71052

daniel-beck commented:
Closing in favor of #8211 which is about to be merged.
